practices/appeng/adrs/0006-use-graphql.md

```diff
@@ -0,0 +1,83 @@
+
+Specifically, with respect to it being a graph, it is very common for mature REST implementations to add features for including related objects in a request for a single object. You might `GET /customers/3?include=[‘address’, ‘shipments.approver’]` for instance, to reduce the number of requests that a client needs to make to render a specific page. That kind of thing is something every team that implements a REST API ends up needing to specify and implement themselves. GraphQL took that, and made it a core part of the protocol. If you have a page that needs those related entities, you just specify that in your request, and the server will return all of them together.
+
+There are many different trade offs involved in an API transport protocol. Message size is an easy one to measure, but others like flexibility, standardization, and prevalence are important as well. In considering choosing a default protocol for Truss, we want to pick something that sets up an application engineering team for success as much as possible.
+
+## Decision
+
+Truss projects should default to using GraphQL for communicating between the client and the server. The strict specification of the protocol allows for tools like [GraphQL Code Generator](https://www.graphql-code-generator.com) to generate very solid code, and saves teams time from having to figure out how to implement those kinds of features in REST from scratch. It eliminates the hassle of building custom endpoints to support specific front end requests except when there is a proven performance need, by default you just design your resources and their relationships. It simplifies client code by allowing most pages to get their data from a single request, reducing the number of error cases the UI needs to handle.
+
+## Why is this Applicable to the Practice as a Whole
+
+Standardizing on an API transport layer opens the door to increasingly sharing tools and patterns across different teams. Since, as a contracting org, we regularly start new projects, having a library of tools and patterns and experience we can rely on makes it easier for teams to get going at the start of a project.
+
+For the API transport layer in particular, this is a decision where there are many different ways to accomplish the same goal, which is essentially to provide a medium for the app to make requests of the API. GraphQL is good enough and the benefits of standardizing outweigh any of the benefits of alternatives.
+
+## When to Not Implement This Decision
+
+- If you need to work with multi-part uploads you will probably need to do that outside of graphQL at the very least.
+- If your project requires very high network performance (or very low network traffic) other tools will give you more control over message size
+- If you need to build an API for external consumption, GraphQL might be too big of an ask for other teams to figure out
+
```

Comment:
We could add something here along the lines of: `If GraphQL library support is not strong in your target tech stack`.  You do note below that GraphQL is more difficult on Rails, but it may be worth a more general call out.